### ISSUE TITLE
Is Template (render) thread safe? 

### ISSUE THREAD

**Issue by zacMode** [bkiers/Liqp#287]
Hi, i'm new to this library (and ANTLR), but looking to add it to a project. I'm trying to understand how a Template  instance should be used. Is it safe to parse a liquid string once and then use the single Template instance concurrently to render different strings? 

e.g. 
```
Template template = Template.parse(" hello, {{ name }}");

final ExecutorService executorService = Executors.newCachedThreadPool()
for (int i  = 0; i < 10_000; i ++) {
  executorService.submit(()-> template.render(Map.of("name", randomName())));
}
```

Thanks!

**Comment by msangel** [comment]
We made a lot of effort in moving forward to that, but still, no. 
Each `render` call redefine internal `templateContext` field, so it is not safe yet to use this in multi-threaded env. 

Keeping this ticket open until it will be thread-safe.

**Comment by msangel** [comment]
Also `errors()` method depends on the context, so we cannot do it local/parameter. it should be field.
![image](https://github.com/bkiers/Liqp/assets/773163/dd4b55ec-9a9a-4665-8236-312507413941)

Using Thread local for that is problematic. (we want know errors as long as it could)

**Comment by arunkoshi** [comment]
Hi @msangel Just wanted to check if there is any work planned towards making the template thread safe?
 
We have a situation where our customers setup liquid templates, and we render those in a spring api service as part of request processing.

We are observing that for the more complex templates, the time to `parse` the liquid template is quite high (of the order of 2 seconds to upto 30 seconds if the server cpu is running busy), although the `render` part happens in few milliseconds.  

We were hoping to cache the parsed template as a solution, but realized that it is not multi-threading safe from this issue. 

For now, we are considering adding a thread local LRU cache to avoid multi-threading issues. But looking at the memory taken by the template object (of order ~1.5 mb per template instance), it looks like we will need to use a very tiny limit of max_cached_templates_per_thread in the LRU cache (since Spring can spin up quite a lot of threads in its pool).

Wondering if there are any recommendations or if we can have any insight on plans to make it multi-threading safe? 

PS: 
1. We are using version liqp 0.9.1.0
2. We were using ruby liquid in our legacy app earlier, which does not seem to take as much time for the parsing.

**Comment by msangel** [comment]
@arunkoshi 
sorry, I paused the work on the project for undefined time. I live in Ukraine and have other stuf to do now.

**Comment by msangel** [comment]
All mine comments above still valid.

**Comment by arunkoshi** [comment]
Thanks @msangel. Appreciate you getting back!

Would you please be able to give some feedback on our approach of using cached ThreadLocal Template instances (so each thread parses and keeps its own copy of the heavy templates)? 

Essentially we execute following snippet while handling a request in Spring, in order to get the rendered template. 

```
  TemplateParser templateParser = new TemplateParser.Builder()
                  .withFlavor(Flavor.LIQUID).build();
  try {
            Template parsedTemplate;
             if(templateKey already in ThreadLocal map) {
                parsedTemplate = getTemplateFromThreadLocalMap(templateKey);
            }
            else {
                parsedTemplate = templateParser.parse(requestTemplate);
                putTemplateInThreadLocalMap(templateKey, parsedTemplate);
            } 
              return parsedTemplate.render(contextObject);
          } catch (LiquidException e) {
              LOGGER.info("Issue with the template " + e.getMessage());
                  throw e;
          }
```
I am wondering if we need to add any additional safeguards like discarding the cached `parsedTemplate` in case of any LiquidException from the `render` method?

PS:
1. We do not use the `errors()` method of the `template` instance.

**Comment by msangel** [comment]
It doesn’t keep any internals after rendering other than errors, also as I remember it should clear errors on new rendering. 

Still, I would probably use another caching pattern— object pool, where you take the template off the caching collection before use and then return back.

In simple way that can be achieve by Map<String name, List<Template> allWithTheName> so you simply look list by key, then take out available from list, if not — create local, and after processing put to list. 

But that might be overwhelming and your approach just works 🙂

**Comment by arunkoshi** [comment]
Yes, object pool was something we discussed internally, but for now we are looking for simplest fix possible.

Really appreciate you responding. Thanks for all your inputs!
